bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java

```diff
@@ -0,0 +1,193 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * Handles the long polling to the Smart Home Controller.
+ * 
+ * @author Christian Oeing - Initial contribution
+ */
+@NonNullByDefault
+public class LongPolling {
+    @SuppressWarnings("serial")
+    private class AbortLongPolling extends BoschSHCException {
+    }
+
+    private final Logger logger = LoggerFactory.getLogger(LongPolling.class);
+
+    private final Gson gson = new Gson();
+
+    private final ScheduledExecutorService scheduler;
+
+    /**
+     * Handler for long poll results.
+     */
+    private final Consumer<LongPollResult> handleResult;
+
+    /**
+     * Handler for unrecoverable.
+     */
+    private final Consumer<Throwable> handleFailure;
+
+    /**
+     * Current running long polling request.
+     */
+    private @Nullable Request request;
+
+    /**
+     * Indicates if long polling was aborted.
+     */
+    private boolean aborted = false;
+
+    public LongPolling(ScheduledExecutorService scheduler, Consumer<LongPollResult> handleResult,
+            Consumer<Throwable> handleFailure) {
+        this.scheduler = scheduler;
+        this.handleResult = handleResult;
+        this.handleFailure = handleFailure;
+    }
+
+    public void start(BoschHttpClient httpClient) throws LongPollingFailedException {
+        // Subscribe to state updates.
+        String subscriptionId = this.subscribe(httpClient);
+        this.executeLongPoll(httpClient, subscriptionId);
+    }
+
+    public void stop() {
+        // Abort long polling.
+        this.aborted = true;
+        Request request = this.request;
+        if (request != null) {
+            request.abort(new AbortLongPolling());
+            this.request = null;
+        }
+    }
+
+    /**
+     * Subscribe to events and store the subscription ID needed for long polling
+     *
+     * Method is synchronous.
+     * 
+     * @return Subscription id
+     */
+    private String subscribe(BoschHttpClient httpClient) throws LongPollingFailedException {
+        try {
+            String url = httpClient.createUrl("remote/json-rpc");
+            JsonRpcRequest request = new JsonRpcRequest("2.0", "RE/subscribe",
+                    new String[] { "com/bosch/sh/remote/*", null });
+            logger.debug("Subscribe: Sending request: {} - using httpClient {}", gson.toJson(request), httpClient);
+            Request httpRequest = httpClient.createRequest(url, POST, request);
+            SubscribeResult response = httpClient.sendRequest(httpRequest, SubscribeResult.class);
+
+            logger.debug("Subscribe: Got subscription ID: {} {}", response.getResult(), response.getJsonrpc());
+            String subscriptionId = response.getResult();
+            return subscriptionId;
+        } catch (TimeoutException | ExecutionException | InterruptedException e) {
+            logger.error("Error on subscribe request", e);
```

Comment:
The error logging level should be reserved for issues catastrophic enough to threaten the operation of openHab itself. An error in a binding shouldn't qualify for this. Please read the [openHab coding guidelines](https://www.openhab.org/docs/developer/guidelines.html#f-logging) for expected use of logging in bindings.

Reply:
Removed the log completely as a warning will be already logged by the BoschBridgeHandler when a LongPollingException occurs.